File: abc/075/c.hs
import qualified Data.Map as Map
import Control.Monad
import Control.Monad.ST
import Data.STRef

main :: IO ()
main = do
  [n, m] <- (map read . words) <$> getLine :: IO [Int]
  (graph, edges) <- buildGraph (Map.fromList [] ) [] m
  print $ length [True | rm <- edges, bfs graph 1 rm < n]

buildGraph graph edges 0 = return (graph, edges)
buildGraph graph edges n = do
  [a, b] <- (map read . words) <$> getLine :: IO [Int]
  let graph' = Map.insertWith (++) a [b] graph
  let graph'' = Map.insertWith (++) b [a] graph'
  buildGraph graph'' ((a, b):edges) (n - 1)

-- あるノードから辿れるノードの総数を求める
bfs :: Map.Map Int [Int] -> Int -> (Int, Int) -> Int
bfs graph start rm = runST $ do
  visitedRef <- newSTRef (Map.fromList [] :: Map.Map Int Bool)
  count <- newSTRef (0 :: Int)
  let queue = [start]
      loop [] = return ()
      loop queue = do
        let p:queue' = queue
        let maybeNeighbors = Map.lookup p graph
        case maybeNeighbors of
          Nothing -> undefined
          Just neighbors ->
            forM_ neighbors $ \n -> do
              visited <- readSTRef visitedRef
              let v = Map.lookup n visited
              case v of
                Just True -> return ()
                Nothing ->
                  if (p == fst rm && n == snd rm) || (p == snd rm && n == fst rm)
                    then return ()
                    else do
                      let visited' = Map.insert n True visited
                      writeSTRef visitedRef visited'
                      modifySTRef count (+1)
                      loop (queue' ++ [n])

  loop queue
  readSTRef count

File: abc/075/c.rb
def bfs(graph, start, rm)
  visited = {}
  count = 0
  queue = [start]
  until queue.empty? do
    p = queue.shift
    neighbors = graph[p]
    neighbors.each do |n|
      next if visited[n]
      next if (p == rm[0] && n == rm[1]) || (p == rm[1] && n == rm[0])
      visited[n] = true
      queue.push(n)
      count += 1
    end
  end
  count
end

@visited = {}
@count = 0
def dfs(graph, now, rm)
  @visited[now] = true
  @count += 1
  neighbors = graph[now]
  neighbors.each do |n|
    next if @visited[n]
    next if (now == rm[0] && n == rm[1]) || (now == rm[1] && n == rm[0])
    dfs(graph, n, rm)
  end
end

n, m = gets.strip.split.map(&:to_i)
graph = {}
(1..n).each do |i|
  graph[i] = []
end

edges = []
m.times do
  a, b = gets.strip.split.map(&:to_i)
  graph[a] << b
  graph[b] << a
  edges << [a, b]
end

num_bridges = 0
edges.each do |e|
  @visited = {}
  @count = 0
  dfs(graph, 1, e)
  #count = bfs(graph, 1, e)
  num_bridges += 1 if @count < n
end

puts num_bridges
